samples/nrf9160/serial_lte_modem/src/slm_at_icmp.c

```diff
@@ -0,0 +1,516 @@
+			goto close_end;
+		}
+		if (len < header_len) {
+			LOG_INF("nrf_recv() wrong data (%d)", len);
+			goto reread;
+		}
+
+		delta_t = k_uptime_delta(&start_t);
+
+		/* Check for ICMP EchoReply */
+		if (data[0] != rep) {
+			/* We got unsolicited message
+			 * drop it and wait for echo-reply
+			 */
+			goto reread;
+		} else {
+			if (ip_buf[IP_PROTOCOL_POS] != ICMP) {
+				/* Not ipv4 echo reply, ignore silently */
+				goto reread;
+			} else {
+				/* Check ICMP HCS */
+				hcs = check_ics(data, len - header_len);
+				if (hcs != 0) {
+					LOG_ERR("HCS error %d", hcs);
+					goto close_end;
+				}
+				/* Payload length */
+				pllen = (ip_buf[2] << 8) + ip_buf[3];
+			}
+
+			/* Check seqnr and length */
+			plseqnr = data[7];
+			if (plseqnr != seqnr) {
+				LOG_ERR("error sequence numbers %d %d",
+					plseqnr, seqnr);
+				goto close_end;
+			}
+			if (pllen != len) {
+				LOG_ERR("error length %d %d", pllen, len);
+				goto close_end;
+			}
+
+			/* FIXME: Check Code */
+
+			/* Result */
+			sprintf(buf, "#XPING: %d.%03d\r\n",
+				(u32_t)(delta_t)/1000,
+				(u32_t)(delta_t)%1000);
+			m_callback(buf);
+		}
+
+close_end:
+		(void)nrf_close(fd);
+	} while (0);
+
+	return (u32_t)delta_t;
+}
+
+static void ping_task(void *arg1, void *arg2, void *arg3)
+{
+	struct addrinfo *si = argv.src;
+	struct addrinfo *di = argv.dest;
+	u32_t sum = 0;
+	u32_t count = 0;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	for (int i = 0; i < argv.count; i++) {
+		u32_t to = send_ping_wait_reply();
+
+		if (to > 0)  {
+			count++;
+			sum += to;
+		}
+		k_sleep(argv.interval);
+	}
+
+	if (count > 1) {
+		u32_t ave = (sum + count/2) / count;
+		int ave_s = ave / 1000;
+		int ave_f = ave % 1000;
+
+		sprintf(buf, "#XPING: average %d.%03d\r\n", ave_s, ave_f);
+		m_callback(buf);
+		m_callback("OK\r\n");
+	}
+
+	freeaddrinfo(si);
+	freeaddrinfo(di);
+	icmp_thread_id = NULL;
+}
+
+static int ping_test_handler(const char *url, int length, int waittime,
+			int count, int interval)
+{
+	int st;
+	struct addrinfo *res;
+	int addr_len;
+
+	st = modem_info_params_get(&modem_param);
+	if (st < 0) {
+		LOG_ERR("Unable to obtain modem parameters (%d)", st);
+		return -1;
+	}
+
+	/* Check network connection status by checking local IP address */
+	addr_len = strlen(modem_param.network.ip_address.value_string);
+	if (addr_len == 0) {
+		LOG_ERR("LTE not connected yet");
+		return -1;
+	}
+	if (!check_for_ipv4(modem_param.network.ip_address.value_string,
+			addr_len)) {
+		LOG_ERR("Invalid local address");
+		return -1;
+	}
+
+	/* Get own IP address */
+	st = getaddrinfo(modem_param.network.ip_address.value_string,
+			NULL, NULL, &res);
+	if (st != 0) {
+		LOG_ERR("getaddrinfo(src) error: %d", st);
+		freeaddrinfo(argv.dest);
+		return -st;
+	}
+	argv.src = res;
+
+	/* Get destination */
+	res = NULL;
+	st = getaddrinfo(url, NULL, NULL, &res);
+	if (st != 0) {
+		LOG_ERR("getaddrinfo(dest) error: %d", st);
```

Comment:
I think we should `freeaddrinfo(argv.src);` here, it looks that the `getaddrinfo` calls were swapped.

Reply:
Yes, exactly. I swapped the call after originally dest first, src second, to src first then dest.